File: src/main/java/factory/creator/VehicleFactory.java
package factory.creator;

import factory.product.Vehicle;
import factory.product.VehicleType;

public abstract class VehicleFactory {
    public Vehicle orderVehicle(VehicleType vehicleType) {
        var vehicle = createVehicle(vehicleType);

        vehicle.assembly();
        vehicle.loadShip();

        return vehicle;
    }

    protected abstract Vehicle createVehicle(VehicleType vehicleType);
}

File: kotlin/src/main/kotlin/factory/creator/VehicleFactory.kt
package factory.creator

import factory.product.Vehicle
import factory.product.VehicleType

abstract class VehicleFactory {

    fun orderVehicle(vehicleType: VehicleType): Vehicle {
        val vehicle = createVehicle(vehicleType)

        vehicle.assembly()
        vehicle.loadShip()

        return vehicle
    }

    protected abstract fun createVehicle(vehicleType: VehicleType): Vehicle
}